protocols/jabber/iq.c

```diff
@@ -723,6 +776,62 @@ xt_status jabber_iq_query_server(struct im_connection *ic, char *jid, char *xmln
 	return jabber_write_packet(ic, query) ? XT_HANDLED : XT_ABORT;
 }
 
+xt_status jabber_iq_parse_gmail(struct im_connection *ic, struct xt_node *node, struct xt_node *orig)
+{
+	struct xt_node *c;
+	struct jabber_data *jd = ic->proto_data;
+	char *xmlns, *from;
+
+	if (!(c = xt_find_node(node->children, "mailbox")) ||
+	    !(from = xt_find_attr(node, "from")) ||
+	    !(xmlns = xt_find_attr(c, "xmlns"))) {
+		imcb_log(ic, "WARNING: Received incomplete mailbox packet for gmail notify");
+		return XT_HANDLED;
+	}
+	if (strcmp(xmlns, XMLNS_GMAILNOTIFY) == 0) {
+		unsigned long long l_time = 0;
+		char *tid = NULL;
+		c = c->children;
+
+		while ((c = xt_find_node(c, "mail-thread-info"))) {
+			char *subject, *snippet, *msg;
+			struct xt_node *thread;
+			unsigned long long t_time;
+
+			t_time = strtoull(xt_find_attr(c, "date"), NULL, 10);
+			if (t_time && t_time > l_time) {
+				l_time = t_time;
+				tid = xt_find_attr(c, "tid");
+			}
+
+			thread = c->children;
+			subject = xt_find_node(thread, "subject")->text;
```

Comment:
this looks pretty and it would be neat to be able to do this safely, but it could dereference a null there if there's no subject